### PR TITLE
Introduce the keystone usb into keplr wallet

### PR DESCRIPTION
**Pull request by aaronisme**
Hi Team, This is PR is for introducing the keystone USB support for keplr extenstion wallet.

### CODE REVIEW

File: apps/extension/src/pages/sign/cosmos/adr36.tsx [hunk 1]
```diff
@@ -108,6 +109,11 @@ export const SignCosmosADR36Page: FunctionComponent = observer(() => {
     Error | undefined
   >(undefined);
 
+  const isKeystonUSB =
```
Comment (Thunnini) [inline apps/extension/src/pages/sign/cosmos/adr36.tsx:112]: I seems that this is typo of `isKeystoneUSB`

File: apps/extension/src/pages/register/name-password-hardware/index.tsx [hunk 2]
```diff
@@ -39,10 +39,12 @@ export const RegisterNamePasswordHardwareScene: FunctionComponent<{
   const form = useFormNamePassword();
 
   const [connectTo, setConnectTo] = useState<string>("Cosmos");
-
-  const bip44PathState = useBIP44PathState(type === "ledger");
+  const bip44PathState = useBIP44PathState(true);
   const [isBIP44CardOpen, setIsBIP44CardOpen] = useState(false);
 
+  const [keystoneWay, setKeystoneWay] = useState<string>("USB");
+  const [isKeystoneUSB, setIsUSB] = useState(true);
```
Comment (Thunnini) [inline apps/extension/src/pages/register/name-password-hardware/index.tsx:46]: I think `setIsUSB` should be `setIsKeystoneUSB`

File: packages/background/src/keyring-keystone/service.ts [hunk 3]
```diff
@@ -31,6 +31,7 @@ export class KeyRingKeystoneService {
         device: multiAccounts.device,
         deviceId: multiAccounts.deviceId,
         keys,
+        connectionType: multiAccounts.connectionType,
```
Comment (Thunnini) [inline packages/background/src/keyring-keystone/service.ts:34]: Please make sure to additionally store bip44Path here. If there is bip44Path under insensitive, the bip44 path will be displayed to the user on the wallet select page. In the case of USB, it seems that the bip44 path can be used, but users often forget the path, so it needs to be displayed. See the implementation of keyring-mnemonic.

**Review by Thunnini — CHANGES_REQUESTED** [review]

**Comment by Thunnini** [comment]
Looking at the code itself, there doesn't seem to be any major problems. However, the problem is that USB connection does not work in my environment. After connecting via USB in register, the USB permission modal in the web browser appears, but there is no response after selecting keystone.

**Comment by aaronisme** [comment]
> Looking at the code itself, there doesn't seem to be any major problems. However, the problem is that USB connection does not work in my environment. After connecting via USB in register, the USB permission modal in the web browser appears, but there is no response after selecting keystone.

Hi @Thunnini This USB function also need some change on the device side, I will share you and your team the beta firmware so you can test this PR.

File: apps/extension/src/pages/register/connect-keystone/usb.tsx [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    let device: string | undefined;
+    let masterFingerprint = "";
+    let deviceId: string | undefined;
+    let accounts;
+
+    try {
+      if (!isDefaultPath(bip44Path)) {
+        const setPath = `m/44'/118'/${bip44Path.account}'/${bip44Path.change}/${bip44Path.addressIndex}`;
+        const res = await baseApp.getURAccount(
+          setPath,
```
[the remark this answers is not in the document]
Comment (aaronisme) [reply apps/extension/src/pages/register/connect-keystone/usb.tsx:122]: The logics here is the new USB integration enables users to choose a custom derivation path when importing from their Keystone device, a feature not offered by the existing QR method. By clicking the advanced button, users can specify their desired path, aligning this process with other hardware connections, such as Ledger.

image
image

[the remark this answers is not in the document]
Comment (aaronisme) [reply apps/extension/src/pages/register/connect-keystone/usb.tsx:122]: If the user does not select the `advanced` connection method, all supported Keystone coin types will be retrieved during the connection process, including the Ethereum path. The available paths are as follows:

```
const DEFAULT_KEYSTONE_PATHS = [
  "m/44'/118'/0'/0/0", // cosmos path
  "m/44'/60'/0'/0/0", // ethereum path
  "m/44'/529'/0'/0/0",
  "m/44'/394'/0'/0/0",
  "m/44'/234'/0'/0/0",
  "m/44'/564'/0'/0/0",
  "m/44'/459'/0'/0/0",
  "m/44'/330'/0'/0/0",
];
```

These paths are aligned with the existing QR connection method.

[the remark this answers is not in the document]
Comment (aaronisme) [reply apps/extension/src/pages/register/connect-keystone/usb.tsx]: @Thunnini I added coin type 60 for users with a custom derivation path.

File: apps/extension/src/pages/sign/cosmos/adr36.tsx (hunk 1 above)
Comment (aaronisme) [reply apps/extension/src/pages/sign/cosmos/adr36.tsx]: fixed

File: apps/extension/src/pages/register/name-password-hardware/index.tsx (hunk 2 above)
Comment (aaronisme) [reply apps/extension/src/pages/register/name-password-hardware/index.tsx]: fixed

File: packages/background/src/keyring-keystone/service.ts (hunk 3 above)
Comment (aaronisme) [reply packages/background/src/keyring-keystone/service.ts:35]: added

**Comment by aaronisme** [comment]
@Thunnini All comment issues have been resolved, please verify. And I have shared the beta firmware in our tg group

**Pull request merged**